src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,858 @@
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, _inputColumnName) };
+        }
+
+        protected override Batch CreateBatch(DataViewRowCursor input) => new Batch(_batchSize, _outputLength, _threshold, _sensitivity, _detectMode);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => _batchSize == -1 ? input.Position == 0 : input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => _batchSize == -1 ? input.Position == -1 : (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override ValueGetter<double> GetLookAheadGetter(DataViewRowCursor input)
+        {
+            return input.GetGetter<double>(input.Schema[_inputColumnName]);
+        }
+
+        protected override Func<int, bool> GetSchemaBindingDependencies(Func<int, bool> predicate)
+        {
+            return _bindings.GetDependencies(predicate);
+        }
+
+        protected override void ProcessExample(Batch currentBatch, double currentInput)
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        internal sealed class Batch
+        {
+            private List<double> _previousBatch;
+            private List<double> _batch;
+            private readonly int _outputLength;
+            private SrCnnEntireModeler _modeler;
+            private int _batchSize;
+            private double[][] _results;
+
+            public Batch(int batchSize, int outputLength, double threshold, double sensitivity, SrCnnDetectMode detectMode)
+            {
+                _batchSize = batchSize;
+                _outputLength = outputLength;
+                if (batchSize == -1)
+                {
+                    _previousBatch = new List<double>();
+                    _batch = new List<double>();
+                }
+                else
+                {
+                    _previousBatch = new List<double>(batchSize);
+                    _batch = new List<double>(batchSize);
+                }
+                _modeler = new SrCnnEntireModeler(threshold, sensitivity, detectMode);
+            }
+
+            public void AddValue(double value)
+            {
+                _batch.Add(value);
+            }
+
+            public int Count => _batch.Count;
+
+            public void Process()
+            {
+                _batchSize = _batch.Count;
+                if (_batch.Count < MinBatchSize)
+                {
+                    if (_previousBatch.Count + _batch.Count < MinBatchSize)
+                        return;
```

Comment:
>return [](start = 24, length = 6)

In what scenario does this happen?
What does it mean for the getter, doesn't it need to return something? It seems that if this happens it will just return the "old" values in `_results`.

Reply:
Changed the condition to _previousBatch.Count == 0 to make it easier to understand. This happens when batchSize is set to -1 but the input point count < 12. I changed the return to throw Exception here.
                    if (_previousBatch.Count == 0)
                    {
                        throw new InvalidOperationException("The input must contain no less than 12 points.");
                    }

---
In reply to: [428515960](https://github.com/dotnet/machinelearning/pull/5135#discussion_r428515960) [](ancestors = 428515960)